bin/cron/applab_datasets/spotify

```diff
@@ -4,100 +4,132 @@ abort 'Script already running' unless only_one_running?(__FILE__)
+  end
+
+  regex = generate_profanity_regex(items)
+  records = {}
   id = 1
-  response.body.force_encoding('utf-8')
-  csv = CSV.new(response.body)
-  lines = csv.read
-  lines.shift # first line is column names
-  regex = generate_profanity_regex(lines)
-  lines.each do |line|
-    next if regex&.match?(get_string_to_check(line)) # skip record if it contains profanity
+
+  items.each do |item|
+    track = item['track']
+    next if track.nil? || regex&.match?(get_string_to_check(track)) # skip track if it is nil or contains profanity
 
     record = OpenStruct.new
     record.id = id
-    record.Position = line[0].to_i
-    record['Track Name'] = line[1]
-    record.Artist = line[2]
-    record.URL = line[3]
+    record.Position = id
+    record['Track Name'] = track['name']
+    record.Artist = track['artists'][0]['name'] # first will be main artist, featured artists included in record name
+    if chart_type === CHART_TYPE[:TOP] then record.Popularity = track['popularity'] end
+    record.URL = track['href']
 
     records[id] = record.to_h.to_json
     id += 1
   end
+
   return records, columns
 end
 
-# Only check artist and song names for profanity
-def get_string_to_check(line)
-  "#{line[2]} - #{line[1]} "
+# Only check artist and track names for profanity
+def get_string_to_check(track)
+  if track.nil?
+    return ""
+  else
+    return "#{track['name']} - #{track['artists'][0]['name']} "
+  end
 end
 
 # Generates a case-insensitive regex of profanities in the entire dataset
-# to match each line against.
-def generate_profanity_regex(lines)
-  str = lines.reduce('') {|acc, line| acc + get_string_to_check(line)}
+# to match each item against.
+def generate_profanity_regex(items)
+  str = items.reduce('') {|acc, item| acc + get_string_to_check(item['track'])}
   profanities = ProfanityFilter.find_potential_profanities(str, PROFANITY_LOCALE) || []
   profanities.empty? ? nil : /(#{profanities.join("|")})/i
 end
 
 def main
+  get_token
   fb = FirebaseHelper.new('shared')
-  records, columns = get_top200_data(TOP_200_US_URL)
-  fb.upload_live_table('Top 200 USA', records, columns)
 
-  records, columns = get_top200_data(TOP_200_WORLD_URL)
-  fb.upload_live_table('Top 200 Worldwide', records, columns)
+  records, columns = get_playlist_data(CHART_TYPE[:TOP], TOP_50_GLOBAL_ID)
+  fb.upload_live_table('Top 50 Worldwide', records, columns)
 
-  records, columns = get_viral50_data(VIRAL_50_US_URL)
-  fb.upload_live_table('Viral 50 USA', records, columns)
+  records, columns = get_playlist_data(CHART_TYPE[:TOP], TOP_50_USA_ID)
+  fb.upload_live_table('Top 50 USA', records, columns)
 
-  records, columns = get_viral50_data(VIRAL_50_WORLD_URL)
+  records, columns = get_playlist_data(CHART_TYPE[:VIRAL], VIRAL_50_GLOBAL_ID)
   fb.upload_live_table('Viral 50 Worldwide', records, columns)
+
+  records, columns = get_playlist_data(CHART_TYPE[:VIRAL], VIRAL_50_USA_ID)
+  fb.upload_live_table('Viral 50 USA', records, columns)
```

Comment:
I haven't run this file locally to actually update these tables yet. Wasn't sure if I should or just wait for them to run after being released. I did confirm that they work as expected with the profanity filter.

Reply:
we should probably do a test run beforehand, but i'm not sure how we'd want to time it. maybe we temporarily disable the cron job, merge this change, take a backup of the current data in firebase (or know where to locate it), then run the job manually at a certain time when we're all online